File: 0542-01-matrix/0542-01-matrix.cs
public class Solution {
    public int[][] UpdateMatrix(int[][] mat) {
        // use bfs, initially, push all nodes with '0' in queue. Store ((i,j), dist) in queue
        if(mat.Length == 1 && mat[0].Length == 1)
        {
            return mat;
        }
        int[][] ans = new int[mat.Length][];
        int[][] visited = new int[mat.Length][];
        Queue<Tuple<int,int,int>> bfs = new();
        for(int i=0; i<mat.Length; i++)
        {
            ans[i] = new int[mat[i].Length];
            visited[i] = new int[mat[i].Length];
            for(int j=0; j<mat[i].Length; j++)
            {
                if(mat[i][j] == 0)
                {
                    bfs.Enqueue(Tuple.Create(i,j,0));
                    visited[i][j] = 1;
                }
            }
        }
        int[] deltaRow = [1, 0, -1, 0];
        int[] deltaCol = [0, 1, 0, -1];
        while(bfs.Count > 0)
        {
            var top = bfs.Dequeue();
            int row = top.Item1;
            int col = top.Item2;
            int dist = top.Item3;
            for(int k=0; k<4; k++)
            {
                int newRow = row + deltaRow[k];
                int newCol = col + deltaCol[k];
                if(newRow < mat.Length && newRow >= 0 && newCol < mat[0].Length && newCol >= 0 && visited[newRow][newCol] == 0)
                {
                    visited[newRow][newCol] = 1;
                    ans[newRow][newCol] = dist+1;
                    bfs.Enqueue(Tuple.Create(newRow,newCol,dist+1));
                }
            }
        }
        return ans;
    }
}

File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        Queue<int[]> q = new LinkedList<>();
        boolean[][] vis = new boolean[m][n];
        int[][] ans = new int[m][n];
        for(int i=0; i<m; i++)
        {
            for(int j=0; j<n; j++)
            {
                if(grid[i][j] == 0)
                {
                    q.add(new int[] { i, j, 0 });
                    vis[i][j] = true;
                }
            }
        }
        int[][] DIRS = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        while(q.size() > 0)
        {
            int[] top = q.poll();
            int r = top[0];
            int c = top[1];
            int d = top[2];
            for(int[] dir : DIRS)
            {
                int nr = r + dir[0];
                int nc = c + dir[1];
                if(nr < 0 || nr == m || nc < 0 || nc == n)
                {
                    continue;
                }
                if(!vis[nr][nc] && grid[nr][nc] == 1)
                {
                    vis[nr][nc] = true;
                    ans[nr][nc] = d+1;
                    q.add(new int[] { nr, nc, d+1 });
                }
            }
        }
        return ans;
    }
}